<a href="https://colab.research.google.com/github/WaleedSharkas/hackerfile/blob/master/Exploring_Ebay_Car_Sales_Data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import files
uploaded = files.upload()

Saving autos.csv to autos (4).csv


# Introduction 

In this project, we will work with a dataset on used cars from eBay Kleinanzeigen, a classifieds section from the Germany eBay website. Although the full dataset consisted of over 370,000 listings, we will only use 50,000 data points. This will allow us to demonstrate how to effectively clean a data set from scratch.

The data dictionary provided with data is as follows:

- dateCrawled - When this ad was first crawled. All field-values are taken 
from this date.
-name - Name of the car.
-seller - Whether the seller is private or a dealer.
-offerType - The type of listing
-price - The price on the ad to sell the car.
-abtest - Whether the listing is included in an A/B test.
-vehicleType - The vehicle Type.
-yearOfRegistration - The year in which which year the car was first registered.
-gearbox - The transmission type.
-powerPS - The power of the car in PS.
-model - The car model name.
-kilometer - How many kilometers the car has driven.
-monthOfRegistration - The month in which which year the car was first registered.
-fuelType - What type of fuel the car uses.
-brand - The brand of the car.
-notRepairedDamage - If the car has a damage which is not yet repaired.
-dateCreated - The date on which the eBay listing was created.
-nrOfPictures - The number of pictures in the ad.
-postalCode - The postal code for the location of the vehicle.
-lastSeenOnline - When the crawler saw this ad last online.

The goal of this project is to clean the data and then use pandas to perform some basic initial analysis of the listings.

In [2]:
import pandas as pd
import numpy as np

In [3]:
autos = pd.read_csv('autos.csv', encoding='Latin-1')
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled  ...             lastSeen
0  2016-03-26 17:47:46  ...  2016-04-06 06:45:54
1  2016-04-04 13:38:56  ...  2016-04-06 14:45:08
2  2016-03-26 18:57:24  ...  2016-04-06 20:15:37
3  2016-03-12 16:58:10  ...  2016-03-15 03:16:28
4  2016-04-01 14:38:50  ...  2016-04-01 14:38:50

[5 rows x 20 columns]

Observations on current data:
- Contains 20 columns, most stored as strings
- Some Columns contain null values
- dateCrawled, LastSeen, price, odometer are stored as strings. Will need to be changed 
- certain coloumns will need to be re-named and converted from camelcase to snakecase


In [4]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [5]:
autos.columns= ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']

autos.head()

date_crawled  ...            last_seen
0  2016-03-26 17:47:46  ...  2016-04-06 06:45:54
1  2016-04-04 13:38:56  ...  2016-04-06 14:45:08
2  2016-03-26 18:57:24  ...  2016-04-06 20:15:37
3  2016-03-12 16:58:10  ...  2016-03-15 03:16:28
4  2016-04-01 14:38:50  ...  2016-04-01 14:38:50

[5 rows x 20 columns]

We cleaned up the columns to make them less confusing and more clean:

-yearOfRegistration to registration_year

-monthOfRegistration to registration_month

-notRepairedDamage to unrepaired_damage

-dateCreated to ad_created

Also changed the columns from camelcase to snakecase due to it being the preferred style for Python.

In [6]:
autos.describe(include='all')

date_crawled         name  ...   postal_code            last_seen
count                 50000        50000  ...  50000.000000                50000
unique                48213        38754  ...           NaN                39481
top     2016-03-12 16:06:22  Ford_Fiesta  ...           NaN  2016-04-07 06:17:27
freq                      3           78  ...           NaN                    8
mean                    NaN          NaN  ...  50813.627300                  NaN
std                     NaN          NaN  ...  25779.747957                  NaN
min                     NaN          NaN  ...   1067.000000                  NaN
25%                     NaN          NaN  ...  30451.000000                  NaN
50%                     NaN          NaN  ...  49577.000000                  NaN
75%                     NaN          NaN  ...  71540.000000                  NaN
max                     NaN          NaN  ...  99998.000000                  NaN

[11 rows x 20 columns]

In [7]:
autos["seller"].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [8]:
autos["offer_type"].value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

In [9]:
autos["num_photos"].value_counts()

0    50000
Name: num_photos, dtype: int64

In [10]:
pd.to_datetime(autos["date_crawled"]).max()
# most recent date an observation was collected.

Timestamp('2016-04-07 14:36:56')

In [11]:
autos.loc[~autos["registration_year"].between(1906, 2016), "registration_year"].count()
# number of cars with registration years prior to 1906 or after 2016

1972

In [12]:
 # the number of cars with power rating less than 1ps or greater than 2000ps
autos.loc[~autos["power_ps"].between(1, 2000), "power_ps"].count()

5518

In [13]:
autos["registration_month"].value_counts()

0     5075
3     5071
6     4368
5     4107
4     4102
7     3949
10    3651
12    3447
9     3389
11    3360
1     3282
8     3191
2     3008
Name: registration_month, dtype: int64

Observations:
- we notice that the seller and offer_type columns are almost exclusively a single value. That means they are good candidates to drop as they won't have useful information for analysis. 
-It also appears that the column nr_of_pictures is all zeros, since the minimum and maximum values in the column are both zero, good candidates to drop as they won't have useful information for analysis. 
- the power_ps column has suspicious amount below 1ps
- spicious minimum value of 0, due to There being no zero month. Could be due to typos

In [14]:
autos = autos.drop(["num_photos", "seller", "offer_type"], axis=1)

In [15]:
autos["price"] = (autos["price"]
                          .str.replace("$","")
                          .str.replace(",","")
                          .astype(int)
                          )
autos["price"].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

In [16]:
autos["odometer"] = (autos["odometer"]
                             .str.replace("km","")
                             .str.replace(",","")
                             .astype(int)
                             )
autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)
autos["odometer_km"].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

In [17]:
autos["odometer_km"].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

In [18]:
print(autos["odometer_km"].nunique())

13


In [19]:
(autos["odometer_km"] >= 100000).sum()

39763

Overall, there doesnt seem to be anything that stands out in the odometer_km data.

In [20]:
autos["price"].nunique()

2357

In [21]:
print(autos["price"].describe())
autos["price"].value_counts().head(15)

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
999      434
750      433
900      420
650      419
850      410
Name: price, dtype: int64

In [22]:
autos["price"].value_counts().sort_index(ascending=True).head(30)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
40       6
45       4
47       1
49       4
50      49
55       2
59       1
60       9
65       5
66       1
Name: price, dtype: int64

In [23]:
autos["price"].value_counts().sort_index(ascending=False).head(30)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
194000      1
190000      1
180000      1
175000      1
169999      1
169000      1
163991      1
163500      1
155000      1
151990      1
Name: price, dtype: int64

The price column has numbers in both the max and min list that do not seem unrealistic. Further breakdown of numbers is needed here.

In [24]:
autos.loc[autos["price"] > 350000, ["name", "model", "brand", "registration_year", "price"]]

name  ...     price
514                     Ford_Focus_Turnier_1.6_16V_Style  ...    999999
2897    Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000  ...  11111111
7814                                         Ferrari_F40  ...   1300000
11137  suche_maserati_3200_gt_Zustand_unwichtig_laufe...  ...  10000000
22947                         Bmw_530d_zum_ausschlachten  ...   1234566
24384                            Schlachte_Golf_3_gt_tdi  ...  11111111
27371                                         Fiat_Punto  ...  12345678
37585                                Volkswagen_Jetta_GT  ...    999990
39377                        Tausche_volvo_v40_gegen_van  ...  12345678
39705                        Tausch_gegen_gleichwertiges  ...  99999999
42221                                  Leasinguebernahme  ...  27322222
43049                                      2_VW_Busse_T3  ...    999999
47598  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...  ...  12345678
47634                                        Ferrari_FXX  ...   3890000

[14 rows x 5 columns]

In [25]:
autos.loc[autos["price"] < 0, ["name", "model", "brand", "registration_year", "price"]]

Empty DataFrame
Columns: [name, model, brand, registration_year, price]
Index: []

In [26]:
autos = autos[autos["price"].between(1,400000)]

Price Observations:
- There are 1,421 cars listed with $0 price, It is only 2% of data. We will wait to see if it needs to be removed. 
- It does seem odd that a ford focus would sell for that high. Not sure if it is worth confirming this information. 
- Allow for prices only in the range of 1-400,00

In [27]:
autos[["date_crawled", "ad_created", "last_seen"]] = autos[["date_crawled", "ad_created", "last_seen"]].apply(pd.to_datetime)

In [28]:
autos[["date_crawled", "ad_created", "last_seen"]].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 48565 entries, 0 to 49999
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   date_crawled  48565 non-null  datetime64[ns]
 1   ad_created    48565 non-null  datetime64[ns]
 2   last_seen     48565 non-null  datetime64[ns]
dtypes: datetime64[ns](3)
memory usage: 1.5 MB


In [29]:
autos[["date_crawled", "ad_created", "last_seen"]].head()

date_crawled ad_created           last_seen
0 2016-03-26 17:47:46 2016-03-26 2016-04-06 06:45:54
1 2016-04-04 13:38:56 2016-04-04 2016-04-06 14:45:08
2 2016-03-26 18:57:24 2016-03-26 2016-04-06 20:15:37
3 2016-03-12 16:58:10 2016-03-12 2016-03-15 03:16:28
4 2016-04-01 14:38:50 2016-04-01 2016-04-01 14:38:50

In [30]:
autos["date_crawled"].dt.floor("D").value_counts(normalize = True).sort_index()

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

In [31]:
autos["date_crawled"].dt.floor("D").value_counts(normalize = True).sort_values()

2016-04-07    0.001400
2016-04-06    0.003171
2016-03-18    0.012911
2016-04-05    0.013096
2016-03-06    0.014043
2016-03-13    0.015670
2016-03-05    0.025327
2016-03-24    0.029342
2016-03-16    0.029610
2016-03-27    0.031092
2016-03-25    0.031607
2016-03-17    0.031628
2016-03-31    0.031834
2016-03-10    0.032184
2016-03-26    0.032204
2016-03-23    0.032225
2016-03-11    0.032575
2016-03-22    0.032987
2016-03-09    0.033090
2016-03-08    0.033296
2016-04-01    0.033687
2016-03-30    0.033687
2016-03-29    0.034099
2016-03-15    0.034284
2016-03-19    0.034778
2016-03-28    0.034860
2016-04-02    0.035478
2016-03-07    0.036014
2016-04-04    0.036487
2016-03-14    0.036549
2016-03-12    0.036920
2016-03-21    0.037373
2016-03-20    0.037887
2016-04-03    0.038608
Name: date_crawled, dtype: float64

In [32]:
autos["date_crawled"].dt.hour.value_counts(normalize = True).sort_index()

0     0.027036
1     0.015196
2     0.006527
3     0.005024
4     0.004118
5     0.003047
6     0.003480
7     0.014084
8     0.019520
9     0.035396
10    0.047812
11    0.055266
12    0.058767
13    0.065109
14    0.068527
15    0.066262
16    0.069742
17    0.070380
18    0.073819
19    0.068012
20    0.068012
21    0.061649
22    0.050839
23    0.042376
Name: date_crawled, dtype: float64

Data Crawled Observation

This data was crawled over daily and around once a month, from March 2016 to april 2016. The overall distribution is faily uniformly scrpaed over. Hourly breakdown shows non uniformly between 2am-6am. 

In [33]:
(autos["last_seen"].dt.floor("D").value_counts(normalize=True).sort_index())

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

Last Seen Observation

This data was crawled over daily and around once a month, from March 2016 to april 2016. Over half of the entries were last seen in the last five days of data.

In [34]:
autos["ad_created"].dt.floor("D").value_counts(normalize = True).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

In [35]:
autos["ad_created"].dt.year.value_counts(normalize = True)

2016    0.999876
2015    0.000124
Name: ad_created, dtype: float64

In [36]:
autos["ad_created"].dt.hour.value_counts(normalize = True).sort_index()

0    1.0
Name: ad_created, dtype: float64

AD Created Observation
 
The ads were created between june 2015  and april 2016. 

In [37]:
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Registratoin Year Observations

There are odd numbers here, it is not possible to have a car in the "1000" which is the min value. The max value is also odd with "9999". Any vehivle registred after 2016 should not be in the data. We should focus on 1900-2016. Further Breakdown needed.

In [38]:
autos.loc[~autos["registration_year"].between(1906, 2016), "registration_year"].value_counts()

2017    1392
2018     470
5000       4
9999       3
2019       2
1800       2
2800       1
6200       1
4500       1
8888       1
4800       1
1000       1
9000       1
1001       1
1111       1
5911       1
4100       1
Name: registration_year, dtype: int64

In [39]:
(~autos["registration_year"].between(1900,2016)).sum() / autos.shape[0]
# Percent of invalid data 

0.038793369710697

In [40]:
autos.loc[autos["registration_year"].between(1906, 2016), "registration_year"].describe()

count    46681.000000
mean      2002.910756
std          7.185103
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

In [41]:
autos.loc[autos["registration_year"].between(1995, 2016), "registration_year"].value_counts(normalize = True).sort_index()

1995    0.028354
1996    0.031728
1997    0.045085
1998    0.054606
1999    0.066946
2000    0.072931
2001    0.060914
2002    0.057448
2003    0.062370
2004    0.062462
2005    0.067847
2006    0.061700
2007    0.052618
2008    0.051185
2009    0.048181
2010    0.036720
2011    0.037505
2012    0.030272
2013    0.018556
2014    0.015321
2015    0.009059
2016    0.028192
Name: registration_year, dtype: float64

In [42]:
autos.loc[autos["registration_year"].between(1906, 2016), "registration_year"].value_counts(normalize = True).sort_index().loc[1970:].sum()

0.9955013817184722

Registratoin Year Observations Final

Due to around 99% of this data occuring after 1970. We will set a date range for this data " 1970 - 2016".

In [43]:
autos = autos[autos["registration_year"].between(1969,2016)]
autos["registration_year"].value_counts(normalize=True).head(30)

2000    0.067886
2005    0.063153
1999    0.062314
2004    0.058142
2003    0.058055
2006    0.057432
2001    0.056700
2002    0.053474
1998    0.050828
2007    0.048978
2008    0.047645
2009    0.044848
1997    0.041966
2011    0.034911
2010    0.034179
1996    0.029533
2012    0.028178
1995    0.026393
2016    0.026242
2013    0.017273
2014    0.014261
1994    0.013530
1993    0.009142
2015    0.008432
1992    0.007959
1990    0.007464
1991    0.007292
1989    0.003743
1988    0.002904
1985    0.002043
Name: registration_year, dtype: float64

In [44]:
autos["brand"].nunique()


40

In [45]:
autos["brand"].value_counts(normalize=True)

volkswagen        0.211831
bmw               0.110389
opel              0.107808
mercedes_benz     0.096365
audi              0.086900
ford              0.069327
renault           0.047279
peugeot           0.029942
fiat              0.025597
seat              0.018327
skoda             0.016434
nissan            0.015337
mazda             0.015251
smart             0.014218
citroen           0.014025
toyota            0.012755
hyundai           0.010067
volvo             0.009056
sonstige_autos    0.008905
mini              0.008798
mitsubishi        0.008260
honda             0.007873
kia               0.007098
alfa_romeo        0.006625
porsche           0.006001
suzuki            0.005958
chevrolet         0.005571
chrysler          0.003485
dacia             0.002646
daihatsu          0.002517
jeep              0.002280
subaru            0.002151
land_rover        0.002065
saab              0.001656
jaguar            0.001527
daewoo            0.001506
rover             0.001334
t

In [46]:
autos["brand"].value_counts(normalize = True).head(10).sum()

0.803764250376425

Brand Observations
- Volkswagen is the most popular brand, with aprox double bmw who is second most popular
- WE will limit our analysis to brands in the top 10 ">0.018"

In [47]:
brand_count = autos["brand"].value_counts(normalize=True)
common_brands = brand_count[brand_count > .018].index
print(common_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat'],
      dtype='object')


In [48]:
brand_mean_prices = {}

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_price = brand_only["price"].mean()
    brand_mean_prices[brand] = int(mean_price)

brand_mean_prices

{'audi': 9331,
 'bmw': 8334,
 'fiat': 2781,
 'ford': 3506,
 'mercedes_benz': 8523,
 'opel': 2973,
 'peugeot': 3088,
 'renault': 2469,
 'seat': 4402,
 'volkswagen': 5391}

In [49]:
autos.groupby("brand")[["odometer_km", "price"]].mean().loc[common_brands].sort_values("price", ascending = False).round(0)

odometer_km   price
audi              129152.0  9332.0
mercedes_benz     130999.0  8524.0
bmw               132667.0  8335.0
volkswagen        128741.0  5391.0
seat              121268.0  4402.0
ford              124963.0  3507.0
peugeot           127209.0  3088.0
opel              129395.0  2973.0
fiat              117408.0  2782.0
renault           128212.0  2470.0

In [50]:
autos.groupby("brand")["model"].value_counts(dropna = False).sort_values(ascending = False).head(10)

brand          model   
volkswagen     golf        3707
bmw            3er         2615
volkswagen     polo        1609
opel           corsa       1591
volkswagen     passat      1349
opel           astra       1348
audi           a4          1231
mercedes_benz  c_klasse    1136
bmw            5er         1132
mercedes_benz  e_klasse     958
Name: model, dtype: int64

In [51]:
autos.groupby("odometer_km")["brand"].describe().sort_index().round(1)

count unique         top  freq
odometer_km                                
5000           752     38  volkswagen   152
10000          237     26  volkswagen    48
20000          734     35  volkswagen   135
30000          755     33  volkswagen   165
40000          791     34  volkswagen   143
50000          982     38  volkswagen   183
60000         1113     39  volkswagen   206
70000         1174     39  volkswagen   206
80000         1365     38  volkswagen   250
90000         1661     39  volkswagen   287
100000        2036     39  volkswagen   368
125000        4847     40  volkswagen   843
150000       30043     40  volkswagen  6862

In [52]:
autos.groupby("brand")["odometer_km"].describe().sort_index().round(1)

count      mean      std  ...       50%       75%       max
brand                                      ...                              
alfa_romeo       308.0  132159.1  33490.7  ...  150000.0  150000.0  150000.0
audi            4040.0  129152.2  38920.1  ...  150000.0  150000.0  150000.0
bmw             5132.0  132666.6  34740.5  ...  150000.0  150000.0  150000.0
chevrolet        259.0   98783.8  46325.9  ...  100000.0  150000.0  150000.0
chrysler         162.0  133086.4  35351.5  ...  150000.0  150000.0  150000.0
citroen          652.0  119677.9  40151.1  ...  150000.0  150000.0  150000.0
dacia            123.0   84268.3  48085.0  ...   80000.0  125000.0  150000.0
daewoo            70.0  121642.9  37837.6  ...  150000.0  150000.0  150000.0
daihatsu         117.0  116410.3  42547.4  ...  125000.0  150000.0  150000.0
fiat            1190.0  117407.6  42307.6  ...  150000.0  150000.0  150000.0
ford            3223.0  124962.8  39244.7  ...  150000.0  150000.0  150000.0
honda            366.0  122172.1  40618.4  ...  150000.0  150000.0  150000.0
hyundai          468.0  106442.3  45290.0  ...  125000.0  150000.0  150000.0
jaguar            71.0  124436.6  40138.6  ...  150000.0  150000.0  150000.0
jeep             106.0  127122.6  38612.5  ...  150000.0  150000.0  150000.0
kia              330.0  112530.3  45149.6  ...  125000.0  150000.0  150000.0
lada              27.0   83518.5  38774.9  ...   80000.0  100000.0  150000.0
lancia            50.0  121900.0  40164.1  ...  150000.0  150000.0  150000.0
land_rover        96.0  118489.6  46843.1  ...  150000.0  150000.0  150000.0
mazda            709.0  124464.0  40037.9  ...  150000.0  150000.0  150000.0
mercedes_benz   4480.0  130998.9  36152.0  ...  150000.0  150000.0  150000.0
mini             409.0   88105.1  43851.3  ...   90000.0  125000.0  150000.0
mitsubishi       384.0  126575.5  39052.8  ...  150000.0  150000.0  150000.0
nissan           713.0  118331.0  41647.3  ...  150000.0  150000.0  150000.0
opel            5012.0  129395.5  37125.1  ...  150000.0  150000.0  150000.0
peugeot         1392.0  127209.1  36031.4  ...  150000.0  150000.0  150000.0
porsche          279.0   97347.7  49284.0  ...  100000.0  150000.0  150000.0
renault         2198.0  128212.0  37013.7  ...  150000.0  150000.0  150000.0
rover             62.0  137661.3  26699.9  ...  150000.0  150000.0  150000.0
saab              77.0  144415.6  22197.9  ...  150000.0  150000.0  150000.0
seat             852.0  121267.6  41222.2  ...  150000.0  150000.0  150000.0
skoda            764.0  110955.5  45782.2  ...  125000.0  150000.0  150000.0
smart            661.0   99326.8  42207.6  ...  100000.0  150000.0  150000.0
sonstige_autos   414.0   91702.9  55682.5  ...  100000.0  150000.0  150000.0
subaru           100.0  125300.0  39111.5  ...  150000.0  150000.0  150000.0
suzuki           277.0  108465.7  43010.5  ...  125000.0  150000.0  150000.0
toyota           593.0  115944.4  41846.5  ...  125000.0  150000.0  150000.0
trabant           55.0   53818.2  36980.5  ...   60000.0   70000.0  150000.0
volkswagen      9848.0  128741.4  38500.2  ...  150000.0  150000.0  150000.0
volvo            421.0  138717.3  28927.6  ...  150000.0  150000.0  150000.0

[40 rows x 8 columns]

In [53]:
autos.groupby("odometer_km")["price"].describe().sort_index().round(1)

count     mean      std  ...      50%      75%       max
odometer_km                             ...                            
5000           752.0   8663.5  26778.2  ...   1500.0   6900.0  350000.0
10000          237.0  20698.6  21146.4  ...  16500.0  27700.0  163991.0
20000          734.0  18355.6  23255.9  ...  14349.5  24200.0  345000.0
30000          755.0  16665.5  16907.3  ...  12800.0  21900.0  220000.0
40000          791.0  15449.0  13661.9  ...  12395.0  19995.0  169999.0
50000          982.0  13679.3  12224.2  ...  10495.0  17950.0  180000.0
60000         1113.0  12330.0  12874.5  ...   9200.0  16000.0  190000.0
70000         1174.0  10866.0   9231.4  ...   8200.0  15000.0   63999.0
80000         1365.0   9710.0   8857.8  ...   7490.0  13000.0  105000.0
90000         1661.0   8387.8   8247.6  ...   5990.0  11400.0   69997.0
100000        2036.0   8022.3  10554.2  ...   5525.0  10599.2  299000.0
125000        4847.0   6205.1   7096.8  ...   4000.0   8450.0  129000.0
150000       30043.0   3754.4   4730.1  ...   2200.0   4800.0  197000.0

[13 rows x 8 columns]

Brand Observations Final

-Mercedes is the most expensive brand, C_klasse top model

-Top 3 brands are Mercedes, audi, and bmw

-Due to Volkswagen having a price that is an avg of the top 10, this 
might be a reason why it is the most sold brand. Top sold model is polo

-The least expensive in the top 10 are fiat and renault

-We can see that the higher milage, the less the price

In [54]:
bmp_series = pd.Series(brand_mean_prices)
pd.DataFrame(bmp_series, columns=["mean_price"])

mean_price
volkswagen           5391
bmw                  8334
opel                 2973
mercedes_benz        8523
audi                 9331
ford                 3506
renault              2469
peugeot              3088
fiat                 2781
seat                 4402

In [55]:

brand_mean_mileage = {}

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_mileage = brand_only["odometer_km"].mean()
    brand_mean_mileage[brand] = int(mean_mileage)

mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)
mean_prices = pd.Series(brand_mean_prices).sort_values(ascending=False)

In [56]:

brand_info = pd.DataFrame(mean_mileage,columns=['mean_mileage'])
brand_info

mean_mileage
bmw                  132666
mercedes_benz        130998
opel                 129395
audi                 129152
volkswagen           128741
renault              128212
peugeot              127209
ford                 124962
seat                 121267
fiat                 117407

In [57]:
brand_info["mean_price"] = mean_prices
brand_info

mean_mileage  mean_price
bmw                  132666        8334
mercedes_benz        130998        8523
opel                 129395        2973
audi                 129152        9331
volkswagen           128741        5391
renault              128212        2469
peugeot              127209        3088
ford                 124962        3506
seat                 121267        4402
fiat                 117407        2781

Price and Milage observation

- Overall the Mileage does not very to much between diffrent brands
- In this list, the more expensive cars have higher prices versus less expensive have much lower prices 

# Conclusion

Overall, we went through some of the steps needed to effectively clean a data set from scratch on the used cars from eBay Kleinanzeigen, a classifieds section from the Germany eBay website. 

Steps taken were as follows:

- Cleaning Column Names
- Simple exploration and Cleaning 
- Exploring certain columns 
- Dealing with Incorrect Registration Dates
- Exploring diffrent comparissons such as price, brand, odomoter 